web/main/utils.py

```diff
@@ -741,3 +742,109 @@ def validate_image(file, formats=None):
+    ...     n.casebook_id = casebook.id
+    ...     n.save()
+    >>> for i, (n, d) in enumerate(zip(nodes[3:], docs[3:])):
+    ...     n.ordinals = (1, 2, i+1)
+    ...     n.resource_id = d.id
+    ...     n.edit_url = ""
+    ...     n.casebook_id = casebook.id
+    ...     n.save()
+    >>> nodes[6].ordinals = (1,)
+    >>> nodes[6].resource_id = docs[6].id
+    >>> nodes[6].save()
+    >>> serialized = manually_serialize_content_query(casebook.contents)
+
+        One top level section, as set up
+    >>> assert len(serialized) == 1
+
+    Serialized data has all expected keys
+        >>> assert all([
+        ...     key in serialized[0].keys()
+        ...     for key in (
+        ...         "title",
+        ...         "id",
+        ...         "edit_url",
+        ...         "url",
+        ...         "citation",
+        ...         "decision_date",
+        ...         "is_transmutable",
+        ...         "ordinals",
+        ...         "ordinal_string",
+        ...         "children",
+        ...     )
+        ... ])
+
+        Serialized data has correct children
+        >>> assert [
+        ...     c["ordinals"] for c in serialized[0]["children"]
+        ... ] == [[1, 1], [1, 2], [1, 3]]
+
+        Serialized data has the correct grandchildren
+        >>> assert [
+        ...     c["ordinals"]
+        ...     for c in serialized[0]["children"][1]["children"]
+        ... ] == [[1, 2, 1], [1, 2, 2], [1, 2, 3]]
+    """
+    from .models import ContentAnnotation, LegalDocument
+    from .serializers import ContentNodeSerializer
+
+    toc = list(
+        content_query.prefetch_resources(
+            legal_doc_query=LegalDocument.objects.defer("content").all()
+        )
+        .order_by("ordinals")
+        .annotate(
+            has_annotation=Exists(ContentAnnotation.objects.filter(resource_id=OuterRef("pk")))
+        )
+        .select_related("casebook")
+        .all()
+    )
+    # optimize expensive call to is_transmutable
+    for t, t1 in zip(toc, toc[1:]):
+        if not t.resource_type or t.resource_type == "Section" or t.resource_type == "":
+            # if t1 is a child of t, then t is not transmutable
+            t.transmutable = t.ordinals != t1.ordinals[: len(t.ordinals)]
+    serialized = {tuple(c.ordinals): ContentNodeSerializer(c).data for c in toc}
+
+    for cns in serialized.values():
+        if cns["resource_type"] == "Section":
+            cns["children"] = []
+
+    serialized[()] = {"children": []}
+
+    for ordinals, cns in serialized.items():
+        if not ordinals:
+            continue
+        try:
+            parent = serialized[ordinals[:-1]]
+        except KeyError:
+            # no parent, append to root
+            parent = serialized[()]
+        try:
+            parent["children"].append(cns)
+        except KeyError:
+            raise ValueError("Trying to append children to non-Section!")
+    return serialized[()]["children"]
```

Comment:
Yes, this is definitely the same --- the story here is that I originally wrote this to work for casebooks only, where all top-level sections have `len(ordinals)==1`. I didn't need a try catch there because the slice alone would take the ordinals to an empty tuple, and this solution felt rather natural. I came back later to modify it for sections, which required the try catch, and required me to give up the nice brevity of the original solution while keeping its extra confusing-ness.